src/journal/catalog.c

```diff
@@ -94,25 +94,85 @@ const struct hash_ops catalog_hash_ops = {
         .compare = catalog_compare_func
 };
 
+static bool next_header(const char **s) {
+        const char *e;
+
+        e = strchr(*s, '\n');
+
+        /* Unexpected end */
+        if (!e)
+                return false;
+
+        /* End of headers */
+        if (e == *s)
+                return false;
+
+        *s = e + 1;
+        return true;
+}
+
+static int combine_entries(const char *one, const char *two, char **result) {
+        const char *b1, *b2;
+        size_t l1, l2, n;
+        char *dest, *p;
+
+        /* Find split point of headers to body */
+        b1 = one;
+        while (next_header(&b1));
+        b2 = two;
+        while (next_header(&b2));
+
+        l1 = strlen(one);
+        l2 = strlen(two);
+        dest = new(char, l1 + l2 + 1);
+        if (!dest)
+                return log_oom();
+
+        p = dest;
+
+        /* Headers from @one */
+        n = b1 - one;
+        memcpy(p, one, n);
+        p += n;
+
+        /* Headers from @two, these will only be found if not present above */
+        n = b2 - two;
+        memcpy(p, two, n);
+        p += n;
+
+        /* Body from @one */
+        n = l1 - (b1 - one);
+        if (n > 0) {
+                memcpy(p, b1, n);
+                p += n;
+
+        /* Body from @two */
+        } else {
+                n = l2 - (b2 - two);
+                memcpy(p, b2, n);
+                p += n;
+        }
+
+        assert(p - dest <= (int)(l1 + l2));
+        p[0] = '\0';
+        *result = dest;
+        return 0;
+}
+
 static int finish_item(
                 Hashmap *h,
-                struct strbuf *sb,
                 sd_id128_t id,
                 const char *language,
-                const char *payload) {
+                char *payload) {
```

Comment:
Oh, duh,yeah the strbuf used to be getting copied to. It's clear, I just hadn't looked closely enough.